protobuf-lite/src/main/java/io/grpc/protobuf/lite/ProtoLiteUtils.java

```diff
@@ -119,7 +121,15 @@ public T parse(InputStream stream) {
         }
         CodedInputStream cis = null;
         try {
-          if (stream instanceof KnownLength) {
+          if (stream instanceof NioBufferBackedStream) {
+            int size = stream.available();
+            if (size > 0) {
+              NioBufferBackedStream nioBufferBackedStream = (NioBufferBackedStream) stream;
+              cis = CodedInputStream.newInstance(nioBufferBackedStream.nioBuffer());
```

Comment:
I think I agree with the principle that both kinds of reading should advance the underlying pointer. I did it the simple way here, as the scope of the optimization was limited to the simple case of a single buffer getting thrown over the fence to the ProtoLite parser, but I'll see if I can get enough of a sense of the whole to be consistent with it.